### PR TITLE
Allow overriding creds paths, and add dummy test command

### PR DESCRIPTION
**Pull request by jonmv**
@mpolden please review. This is just a draft for now.

### CODE REVIEW

**Comment by jonmv** [comment]
@hakonhall FYI

**Comment by jonmv** [comment]
The plan is to pre-populate the `urlsByCluster` of the `cloudTarget` with given `--endpoints`, and if this is done, we skip the wait-for-endpoints when getting a `Service` later down the road.

File: client/go/cmd/test.go [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+var (
+	endpointsArg     string
+	dataPlaneCertArg string
+	dataPlaneKeyArg  string
+)
+
+func init() {
+	rootCmd.AddCommand(testCmd)
+	testCmd.PersistentFlags().StringVar(&endpointsArg, endpointsFlag, "", "The endpoints to use, for each container cluster")
```
Comment (mpolden) [inline client/go/cmd/test.go:30]: Not sure I understand why this is an option. Shouldn't this discover relevant endpoints automatically?

Comment (jonmv) [reply client/go/cmd/test.go:30]: When run by a developer, yes.
When run in the tester-container, endpoints are provided as config, and we can't ask the controller. 

I wonder, though, if these overrides all should be ENV instead?

Comment (mpolden) [reply client/go/cmd/test.go:30]: > When run in the tester-container, endpoints are provided as config, and we can't ask the controller.

Right.

> I wonder, though, if these overrides all should be ENV instead?

Good idea. That what we use for overriding system today, i.e. things regular users don't typically care about.

**Comment by jonmv** [comment]
Code seems to work, but will make some adjustments to output, etc.. Will also add a unit test for the new command, but I think it's ready for review, @mpolden :)

@hakonhall FYI.

File: client/go/cmd/test.go [hunk 2]
```diff
@@ -0,0 +1,357 @@
...
+	Args:              cobra.MaximumNArgs(1),
+	DisableAutoGenTag: true,
+	Run: func(cmd *cobra.Command, args []string) {
+		target := getTarget()
+		testPath := "."
+		if len(args) > 0 {
+			testPath = args[0]
+		}
+		if count, failed := runTests(testPath, target); len(failed) != 0 {
+			_, _ = fmt.Fprintf(stdout, "\nFailed %d of %d tests:\n", len(failed), count)
```
Comment (mpolden) [inline client/go/cmd/test.go:49]: Remove the unnecessary `_, _ =`. Same for all the other occurrences below.

File: client/go/vespa/target.go [hunk 3]
```diff
@@ -129,7 +128,7 @@ func (s *Service) Description() string {
 
 func (t *customTarget) Type() string { return t.targetType }
 
-func (t *customTarget) Service(name string, timeout time.Duration, sessionID int64) (*Service, error) {
+func (t *customTarget) Service(name string, timeout time.Duration, _ int64, _ string) (*Service, error) {
```
Comment (mpolden) [inline client/go/vespa/target.go:131]: I think it's better to keep the parameter names the same as the interface declaration, even if they're unused.

File: client/go/cmd/test.go [hunk 4]
```diff
@@ -0,0 +1,357 @@
...
+		} else {
+			_, _ = fmt.Fprintf(stdout, "%d tests completed successfully\n", count)
+		}
+	},
+}
+
+func runTests(rootPath string, target vespa.Target) (int, []string) {
+	count := 0
+	failed := make([]string, 0)
+	if stat, err := os.Stat(rootPath); err != nil {
```
Comment (mpolden) [inline client/go/cmd/test.go:66]: Not needed? I'm sure `os.ReadDir` fails if `rootPath` is not a directory.

File: client/go/cmd/test.go [hunk 5]
```diff
@@ -0,0 +1,357 @@
...
+func init() {
+	rootCmd.AddCommand(testCmd)
+}
+
+var testCmd = &cobra.Command{
+	Use:   "test [tests directory or test file]",
+	Short: "Run a test suite, or a single test",
+	Long: `Run a test suite, or a single test
+
+Runs all JSON test files in the specified directory, or the single JSON
```
Comment (mpolden) [inline client/go/cmd/test.go:34]: I assume we'll eventually have a documentation page for this feature. We should link to that here when it's added.

File: client/go/cmd/test.go [hunk 6]
```diff
@@ -0,0 +1,357 @@
...
+
+func getParameters(parametersRaw []byte, testsPath string) (map[string]string, error) {
+	if parametersRaw != nil {
+		var parametersPath string
+		if err := json.Unmarshal(parametersRaw, &parametersPath); err == nil {
+			parametersRaw = readFile(path.Join(testsPath, parametersPath))
+		}
+		var parameters map[string]string
+		if err := json.Unmarshal(parametersRaw, &parameters); err != nil {
+			return nil, fmt.Errorf("request parameters must be JSON with only string values: %v", err)
```
Comment (mpolden) [inline client/go/cmd/test.go:298]: Nit: Prefer using `%w` as the format specifier for errors as this allows unwrapping.

File: client/go/cmd/test.go [hunk 7]
```diff
@@ -0,0 +1,357 @@
...
+
+func getBody(bodyRaw []byte, testsPath string) ([]byte, error) {
+	var bodyPath string
+	if err := json.Unmarshal(bodyRaw, &bodyPath); err == nil {
+		return readFile(path.Join(testsPath, bodyPath)), nil
+	}
+	return bodyRaw, nil
+}
+
+func readFile(filePath string) []byte {
```
Comment (mpolden) [inline client/go/cmd/test.go:313]: Use `ioutil.ReadFile` instead, which does the same as this.

File: client/go/cmd/test.go [hunk 8]
```diff
@@ -0,0 +1,357 @@
...
+	failed := make([]string, 0)
+	if stat, err := os.Stat(rootPath); err != nil {
+		fatalErr(err, "Failed reading specified test path")
+	} else if stat.IsDir() {
+		tests, err := os.ReadDir(rootPath)
+		if err != nil {
+			fatalErr(err, "Failed reading specified test directory")
+		}
+		for _, test := range tests {
+			if !test.IsDir() && strings.HasSuffix(test.Name(), ".json") {
```
Comment (mpolden) [inline client/go/cmd/test.go:74]: Nit: Consider `filepath.Ext(test.Name()) == ".json"`.

File: client/go/cmd/test.go (hunk 4 above)
Comment (mpolden) [reply client/go/cmd/test.go:66]: Never mind, I saw the other usage of `stat`.

File: client/go/cmd/test.go (hunk 2 above)
Comment (jonmv) [reply client/go/cmd/test.go:49]: IntelliJ complains when these aren't there, but ok :)

File: client/go/cmd/test.go (hunk 5 above)
Comment (jonmv) [reply client/go/cmd/test.go:34]: Good idea, will add a TODO for this.

File: client/go/vespa/target.go (hunk 3 above)
Comment (jonmv) [reply client/go/vespa/target.go:131]: OK.

File: client/go/cmd/test.go (hunk 6 above)
Comment (jonmv) [reply client/go/cmd/test.go:298]: Thanks!

File: client/go/cmd/test.go (hunk 7 above)
Comment (jonmv) [reply client/go/cmd/test.go:313]: Thanks!

File: client/go/cmd/test.go (hunk 8 above)
Comment (jonmv) [reply client/go/cmd/test.go:74]: Nice.

**Review by mpolden — COMMENTED** [review]
LGTM, but build fails on formatting (can be fixed with `make fmt`)

**Comment by jonmv** [comment]
Yeah, there's a unit tests coming too :)

**Comment by jonmv** [comment]
There we go. PTAL :)

**Comment by jonmv** [comment]
@hakonhall we should probably have another `go` at formatting, names, etc.. But having this merged for testing purposes would be great :)

**Review by mpolden — APPROVED** [review]
👍 build still fails though, but not on Go code.

**Pull request merged**